lib/internal/http/static.js

```diff
@@ -0,0 +1,150 @@
+'use strict';
+
+const {
+  Promise,
+  RegExpPrototypeExec,
+  StringPrototypeEndsWith,
+  StringPrototypeIndexOf,
+  StringPrototypeLastIndexOf,
+  StringPrototypeSlice,
+  StringPrototypeStartsWith,
+  StringPrototypeToLocaleLowerCase,
+} = primordials;
+
+const { cwd } = process;
+const console = require('internal/console/global');
+const { createReadStream } = require('fs');
+const { Server } = require('_http_server');
+const { URL, formatSymbol, isURLInstance, pathToFileURL } = require('internal/url');
+const { emitExperimentalWarning, kEmptyObject } = require('internal/util');
+const { validateFunction, validatePort } = require('internal/validators');
+
+const dot = /\/(\.|%2[eE])/;
+function filterDotFiles(url, fileURL) {
+  return RegExpPrototypeExec(dot, fileURL.pathname) === null;
+}
+const urlFormatOptions = { auth: false, search: false, fragment: false };
+
+function createStaticServer(options = kEmptyObject) {
+  emitExperimentalWarning('http/static');
+
+  const {
+    directory = cwd(),
+    port,
+    host = 'localhost',
+    mimeOverrides,
+    filter = filterDotFiles,
+    log = console.log,
+    onStart = (host, port) => console.log(`Server started on http://${host}:${port}`),
+  } = options;
+  const directoryURL = isURLInstance(directory) ? directory : pathToFileURL(directory);
+  // To be used as a base URL, it is necessary that the URL ends with a slash:
+  const baseDirectoryURL = StringPrototypeEndsWith(directoryURL.pathname, '/') ?
+    directoryURL : new URL(`${directoryURL}/`);
+
+  if (port != null) validatePort(port);
+  if (filter != null) validateFunction(filter, 'options.filter');
+
+  const server = new Server(async (req, res) => {
+    const url = new URL(
+      req.url === '/' || StringPrototypeStartsWith(req.url, '/?') ?
+        './index.html' :
+        '.' + StringPrototypeSlice(new URL(req.url, 'root://')[formatSymbol](urlFormatOptions), 6),
+      baseDirectoryURL,
+    );
+
+    if (filter != null && !filter(req.url, url)) {
+      log?.(403, req.url);
+      res.statusCode = 403;
+      res.end('Forbidden\n');
+      return;
+    }
+
+    const mime = {
+      '__proto__': null,
+      '.html': 'text/html; charset=UTF-8',
+      '.js': 'text/javascript; charset=UTF-8',
+      '.css': 'text/css; charset=UTF-8',
+      '.avif': 'image/avif',
+      '.svg': 'image/svg+xml',
+      '.json': 'application/json',
+      '.jpg': 'image/jpeg',
+      '.jpeg': 'image/jpeg',
+      '.png': 'image/png',
+      '.wasm': 'application/wasm',
+      '.webp': 'image/webp',
+      '.woff2': 'font/woff2',
+      ...mimeOverrides,
+    };
+    const ext = StringPrototypeToLocaleLowerCase(
+      StringPrototypeSlice(url.pathname, StringPrototypeLastIndexOf(url.pathname, '.')),
+    );
+    if (ext in mime) res.setHeader('Content-Type', mime[ext]);
```

Comment:
```suggestion
    const ext = StringPrototypeSlice(url.pathname, StringPrototypeLastIndexOf(url.pathname, '.'))
    const mimeType = mime[ext] || mime[StringPrototypeToLocaleLowerCase(ext)]
    if (mimeType) res.setHeader('Content-Type', mimeType)
```

And maybe we can skip the `StringPrototypeSlice` if the `LastIndexOf` returns -1

Reply:
> Is it to avoid calling 

Yes, is very unlikely the extension will be in uppercase, so a fast-path for lower case could be better.

>  more readable to me as is.

That's okay, I will mark this as resolved, just wanted to push things that could be a little faster.